execution_engine/src/runtime/auction_internal.rs

```diff
@@ -261,65 +310,77 @@ impl<'a, R> MintProvider for Runtime<'a, R>
-                .map_err(|_| Error::Transfer)?
-                .map_err(|_| Error::Transfer)?;
-                return Ok(());
+    fn unbond(&mut self, unbond_kind: &UnbondKind, unbond_era: &UnbondEra) -> Result<(), Error> {
+        let is_delegator = unbond_kind.is_delegator();
+        let (purse, maybe_account_hash) = match unbond_kind {
+            UnbondKind::Validator(pk) | UnbondKind::DelegatedPublicKey(pk) => {
+                let account_hash = pk.to_account_hash();
+                let maybe_value = self
+                    .context
+                    .read_gs_unsafe(&Key::Account(account_hash))
+                    .map_err(|exec_error| {
+                        error!("MintProvider::unbond: {:?}", exec_error);
+                        <Option<Error>>::from(exec_error).unwrap_or(Error::Storage)
+                    })?;
+
+                match maybe_value {
+                    Some(StoredValue::Account(account)) => {
+                        (account.main_purse(), Some(account_hash))
+                    }
+                    Some(StoredValue::CLValue(cl_value)) => {
+                        let entity_key: Key = cl_value.into_t().map_err(|_| Error::CLValue)?;
+                        match self.context.read_gs_unsafe(&entity_key) {
+                            Ok(Some(StoredValue::AddressableEntity(entity))) => {
+                                (entity.main_purse(), Some(account_hash))
+                            }
+                            Ok(Some(StoredValue::CLValue(_))) => {
+                                return Err(Error::CLValue);
+                            }
+                            Ok(Some(_)) => {
+                                return if is_delegator {
+                                    Err(Error::DelegatorNotFound)
+                                } else {
+                                    Err(Error::ValidatorNotFound)
+                                }
+                            }
+                            Ok(None) => {
+                                return Err(Error::InvalidPublicKey);
+                            }
+                            Err(exec_error) => {
+                                error!("MintProvider::unbond: {:?}", exec_error);
+                                return Err(
+                                    <Option<Error>>::from(exec_error).unwrap_or(Error::Storage)
+                                );
+                            }
+                        }
+                    }
+                    Some(_) => return Err(Error::UnexpectedStoredValueVariant),
+                    None => return Err(Error::InvalidPublicKey),
+                }
             }
-            Some(StoredValue::CLValue(cl_value)) => {
-                let contract_key: Key = cl_value.into_t().map_err(|_| Error::CLValue)?;
-                contract_key
+            UnbondKind::DelegatedPurse(addr) => {
+                let purse = URef::new(*addr, AccessRights::READ_ADD_WRITE);
```

Comment:
Seems like the purse is being used to do a read only, is there a reason we are creating it with all the perms here??